Comparison operators in TypeSafeIndex trigger GCC 9 sign-compare warning

For example, from https://drake-jenkins.csail.mit.edu/job/linux-focal-unprovisioned-gcc-bazel-experimental-debug/1/:
```
ERROR: manipulation/kuka_iiwa/BUILD.bazel:35:1: Couldn't build file manipulation/kuka_iiwa/_objs/iiwa_command_receiver/iiwa_command_receiver.pic.o: C++ compilation of rule '//manipulation/kuka_iiwa:iiwa_command_receiver' failed (Exit 1) gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g '-std=c++0x' -MD -MF ... (remaining 205 argument(s) skipped)
In file included from bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/framework_common/drake/systems/framework/framework_common.h:13,
                 from bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/continuous_state/drake/systems/framework/continuous_state.h:15,
                 from bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/leaf_system/drake/systems/framework/leaf_system.h:22,
                 from bazel-out/k8-dbg/bin/manipulation/kuka_iiwa/_virtual_includes/iiwa_command_receiver/drake/manipulation/kuka_iiwa/iiwa_command_receiver.h:12,
                 from manipulation/kuka_iiwa/iiwa_command_receiver.cc:1:
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator==(const U&) const [with U = long unsigned int; Tag = drake::systems::DiscreteStateTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:905:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:351:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
  351 |     return value <= std::numeric_limits<int>::max() &&
      |            ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator==(const U&) const [with U = long unsigned int; Tag = drake::systems::AbstractStateTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:919:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:351:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator==(const U&) const [with U = long unsigned int; Tag = drake::systems::NumericParameterTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:931:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:351:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator==(const U&) const [with U = long unsigned int; Tag = drake::systems::AbstractParameterTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:945:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:351:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator<(const U&) const [with U = long unsigned int; Tag = drake::systems::DiscreteStateTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:1166:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:395:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
  395 |     return value > std::numeric_limits<int>::max() ||
      |            ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator<(const U&) const [with U = long unsigned int; Tag = drake::systems::AbstractStateTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:1172:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:395:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator<(const U&) const [with U = long unsigned int; Tag = drake::systems::NumericParameterTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:1178:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:395:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h: In instantiation of 'typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type drake::TypeSafeIndex<Tag>::operator<(const U&) const [with U = long unsigned int; Tag = drake::systems::AbstractParameterTag; typename std::enable_if<(std::is_integral<U>::value && std::is_unsigned<U>::value), bool>::type = bool]':
bazel-out/k8-dbg/bin/systems/framework/_virtual_includes/system_base/drake/systems/framework/system_base.h:1184:5:   required from here
bazel-out/k8-dbg/bin/common/_virtual_includes/type_safe_index/drake/common/type_safe_index.h:395:18: error: comparison of integer expressions of different signedness: 'const long unsigned int' and 'int' [-Werror=sign-compare]
```

Relates #13102. FYI @jwnimmer-tri @SeanCurtis-TRI.

--- a/common/test/type_safe_index_test.cc
+++ b/common/test/type_safe_index_test.cc
@@ -503,6 +503,67 @@ GTEST_TEST(IntegralComparisons, CompareSizeT) {
       AIndex junk(big_overflow_value), std::runtime_error,
       "Explicitly constructing an invalid index. Type .* has an invalid "
           "value; it must lie in the range .*");
+
+  // Now we test right out at the limit of the *smaller* type -- the int.
+  const AIndex biggest_index{std::numeric_limits<int>::max()};
+  const size_t equal(static_cast<int>(biggest_index));
+  const size_t bigger = equal + 1;
+  const size_t smaller = equal - 1;
+
+  EXPECT_FALSE(biggest_index == smaller);
+  EXPECT_TRUE(biggest_index != smaller);
+  EXPECT_FALSE(biggest_index < smaller);
+  EXPECT_FALSE(biggest_index <= smaller);
+  EXPECT_TRUE(biggest_index > smaller);
+  EXPECT_TRUE(biggest_index >= smaller);
+
+  EXPECT_TRUE(biggest_index == equal);
+  EXPECT_FALSE(biggest_index != equal);
+  EXPECT_FALSE(biggest_index < equal);
+  EXPECT_TRUE(biggest_index <= equal);
+  EXPECT_FALSE(biggest_index > equal);
+  EXPECT_TRUE(biggest_index >= equal);
+
+  EXPECT_FALSE(biggest_index == bigger);
+  EXPECT_TRUE(biggest_index != bigger);
+  EXPECT_TRUE(biggest_index < bigger);
+  EXPECT_TRUE(biggest_index <= bigger);
+  EXPECT_FALSE(biggest_index > bigger);
+  EXPECT_FALSE(biggest_index >= bigger);
+}
+
+// Confirms that comparisons with unsigned types that have fewer bits than
+// TypeSafeIndex's underlying int report propertly.
+GTEST_TEST(TypeSafeIndex, CompareUnsignedShort) {
+  TestScalarComparisons<uint16_t>();
+  TestScalarIncrement<uint16_t>();
+
+  // Test right out at the limit of the *smaller* type -- the uint16_t.
+  const uint16_t big_unsigned = std::numeric_limits<uint16_t>::max();
+  const AIndex bigger_index{static_cast<int>(big_unsigned) + 1};
+  const AIndex smaller_index{static_cast<int>(big_unsigned) - 1};
+  const AIndex equal_index(static_cast<int>(big_unsigned));
+
+  EXPECT_FALSE(bigger_index == big_unsigned);
+  EXPECT_TRUE(bigger_index != big_unsigned);
+  EXPECT_FALSE(bigger_index < big_unsigned);
+  EXPECT_FALSE(bigger_index <= big_unsigned);
+  EXPECT_TRUE(bigger_index > big_unsigned);
+  EXPECT_TRUE(bigger_index >= big_unsigned);
+
+  EXPECT_TRUE(equal_index == big_unsigned);
+  EXPECT_FALSE(equal_index != big_unsigned);
+  EXPECT_FALSE(equal_index < big_unsigned);
+  EXPECT_TRUE(equal_index <= big_unsigned);
+  EXPECT_FALSE(equal_index > big_unsigned);
+  EXPECT_TRUE(equal_index >= big_unsigned);
+
+  EXPECT_FALSE(smaller_index == big_unsigned);
+  EXPECT_TRUE(smaller_index != big_unsigned);
+  EXPECT_TRUE(smaller_index < big_unsigned);
+  EXPECT_TRUE(smaller_index <= big_unsigned);
+  EXPECT_FALSE(smaller_index > big_unsigned);
+  EXPECT_FALSE(smaller_index >= big_unsigned);
 }
 
 // This tests that one index cannot be *constructed* from another index type,

--- a/common/type_safe_index.h
+++ b/common/type_safe_index.h
@@ -348,7 +348,7 @@ class TypeSafeIndex {
       std::is_integral<U>::value && std::is_unsigned<U>::value, bool>::type
   operator==(const U& value) const {
     DRAKE_ASSERT_VOID(AssertValid(index_, "Testing == with invalid index."));
-    return value <= std::numeric_limits<int>::max() &&
+    return value <= static_cast<U>(kMaxIndex) &&
         index_ == static_cast<int>(value);
   }
 
@@ -370,7 +370,7 @@ class TypeSafeIndex {
       std::is_integral<U>::value && std::is_unsigned<U>::value, bool>::type
   operator!=(const U& value) const {
     DRAKE_ASSERT_VOID(AssertValid(index_, "Testing != with invalid index."));
-    return value > std::numeric_limits<int>::max() ||
+    return value > static_cast<U>(kMaxIndex) ||
         index_ != static_cast<int>(value);
   }
 
@@ -392,7 +392,7 @@ class TypeSafeIndex {
       std::is_integral<U>::value && std::is_unsigned<U>::value, bool>::type
   operator<(const U& value) const {
     DRAKE_ASSERT_VOID(AssertValid(index_, "Testing < with invalid index."));
-    return value > std::numeric_limits<int>::max() ||
+    return value > static_cast<U>(kMaxIndex) ||
         index_ < static_cast<int>(value);
   }
 
@@ -414,7 +414,7 @@ class TypeSafeIndex {
       std::is_integral<U>::value && std::is_unsigned<U>::value, bool>::type
   operator<=(const U& value) const {
     DRAKE_ASSERT_VOID(AssertValid(index_, "Testing <= with invalid index."));
-    return value > std::numeric_limits<int>::max() ||
+    return value > static_cast<U>(kMaxIndex) ||
         index_ <= static_cast<int>(value);
   }
 
@@ -436,8 +436,8 @@ class TypeSafeIndex {
       std::is_integral<U>::value && std::is_unsigned<U>::value, bool>::type
   operator>(const U& value) const {
     DRAKE_ASSERT_VOID(AssertValid(index_, "Testing > with invalid index."));
-    return value < std::numeric_limits<int>::max() &&
-        index_ > static_cast<int>(value);
+    return value <= static_cast<U>(kMaxIndex) &&
+           index_ > static_cast<int>(value);
   }
 
   /// Blacklist greater than test with indices of other tags.
@@ -458,8 +458,8 @@ class TypeSafeIndex {
       std::is_integral<U>::value && std::is_unsigned<U>::value, bool>::type
   operator>=(const U& value) const {
     DRAKE_ASSERT_VOID(AssertValid(index_, "Testing >= with invalid index."));
-    return value <= std::numeric_limits<int>::max() &&
-        index_ >= static_cast<int>(value);
+    return value <= static_cast<U>(kMaxIndex) &&
+           index_ >= static_cast<int>(value);
   }
 
   /// Blacklist greater than or equals test with indices of other tags.
@@ -483,7 +483,7 @@ class TypeSafeIndex {
   // Checks if this index lies in the valid range; throws an exception if not.
   // Invocations provide a string explaining the origin of the bad value.
   static void AssertValid(int64_t index, const char* source) {
-    if (index < 0 || index > std::numeric_limits<int>::max()) {
+    if (index < 0 || index > kMaxIndex) {
       throw std::runtime_error(
           std::string(source) + " Type \"" +
           drake::NiceTypeName::Get<TypeSafeIndex<Tag>>() +
@@ -494,7 +494,7 @@ class TypeSafeIndex {
   // This tests for overflow conditions based on adding the given delta into
   // the current index value.
   void AssertNoOverflow(int delta, const char* source) const {
-    if (delta > 0 && index_ > std::numeric_limits<int>::max() - delta) {
+    if (delta > 0 && index_ > kMaxIndex - delta) {
       throw std::runtime_error(
           std::string(source) + " Type \"" +
           drake::NiceTypeName::Get<TypeSafeIndex<Tag>>() +
@@ -509,6 +509,16 @@ class TypeSafeIndex {
   };
 
   int index_{kDefaultInvalid};
+
+  // The largest representable index.
+  // Note: The handling of comparisons of TypeSafeIndex with unsigned integral
+  // types with *fewer* bits relies on truncations of *this* value consisting
+  // of all 1s. The maximum int satisfies that requirement. If, for whatever
+  // reason, some *alternative* maximum index is preferred (or the underlying
+  // integral type of TypeSafeIndex changes), keep this requirement in mind.
+  // Otherwise, comparisons against smaller unsigned integral types is likely
+  // to fail.
+  static constexpr int kMaxIndex = std::numeric_limits<int>::max();
 };
 
 template <typename Tag, typename U>

